src/execution/execute.ts

```diff
@@ -826,32 +825,73 @@ function completeAbstractValue(
 
   return completeObjectValue(
     exeContext,
-    ensureValidRuntimeType(
-      runtimeType,
-      exeContext,
-      returnType,
-      fieldNodes,
-      info,
-      result,
-    ),
+    runtimeType,
     fieldNodes,
     info,
     path,
     result,
   );
 }
 
+function resolveType(
+  exeContext: ExecutionContext,
+  returnType: GraphQLAbstractType,
+  abstractType: GraphQLAbstractType,
+  fieldNodes: ReadonlyArray<FieldNode>,
+  info: GraphQLResolveInfo,
+  result: unknown,
+  encounteredTypeNames: Set<string> = new Set(),
+): GraphQLObjectType | Promise<GraphQLObjectType> {
+  const resolveTypeFn = abstractType.resolveType ?? exeContext.typeResolver;
+  const contextValue = exeContext.contextValue;
+  const possibleRuntimeType = resolveTypeFn(
+    result,
+    contextValue,
+    info,
+    abstractType,
+  );
+
+  if (isPromise(possibleRuntimeType)) {
+    return possibleRuntimeType.then((resolvedPossibleRuntimeType) =>
+      ensureValidRuntimeType(
+        resolvedPossibleRuntimeType,
+        exeContext,
+        returnType,
+        abstractType,
+        fieldNodes,
+        info,
+        result,
+        encounteredTypeNames,
+      ),
+    );
+  }
+  return ensureValidRuntimeType(
+    possibleRuntimeType,
+    exeContext,
+    returnType,
+    abstractType,
+    fieldNodes,
+    info,
+    result,
+    encounteredTypeNames,
+  );
+}
+
 function ensureValidRuntimeType(
```

Comment:
this function is called "ensure" (and previously it just validated type) but now actually recursively resolves type. Please change its name?

Reply:
Sure => what do you suggest? I would like something that implies ensure + recursion as necessary. I guess I thought ensure was good enough, but if you have some candidates that you would prefer, that would be fine.

Reply:
I gave it a go => I was more explicit that we start off with the type name and changed the function to `deriveRuntimeType`.

6c63b4f2